### PR TITLE
Added support for priority queues, RabbitMq headers and mandatory del…

### PR DESCRIPTION
**Pull request by K3llr**
Added support for:

- Priority queues
- RabbitMQ headers
- Mandatory delivery

 __I hereby acknowledge that the code submitted in the pull request has the MIT license and can be merged with the Rebus codebase__.

### CODE REVIEW

**Comment by K3llr** [comment]
RabbitMqCallbackOptionsBuilder is not CLS-compliant due to the events from RabbitMq. This is because BasicReturnEventArgs has a ReplyCode of type ushort. In order to remedy this, we'll have to map the RabbitMq event argument to our oven event type. This might make sense, since we can expose the highlevel message instead (TransportMessage). What do you think would be he right way to do this?

`public class BasicReturnEventArgs : EventArgs
  {
    /// <summary>The content header of the message.</summary>
    public IBasicProperties BasicProperties { get; set; }

    /// <summary>The message body.</summary>
    public byte[] Body { get; set; }

    /// <summary>The exchange the returned message was originally
    /// published to.</summary>
    public string Exchange { get; set; }

    /// <summary>The AMQP reason code for the return. See
    /// RabbitMQ.Client.Framing.*.Constants.</summary>
    public ushort ReplyCode { get; set; }

    /// <summary>Human-readable text from the broker describing the
    /// reason for the return.</summary>
    public string ReplyText { get; set; }

    /// <summary>The routing key used when the message was
    /// originally published.</summary>
    public string RoutingKey { get; set; }
  }`

File: Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs [hunk 1]
```diff
@@ -170,18 +189,28 @@ void BindInputQueue(string address, IModel model)
             model.QueueBind(address, _directExchangeName, address);
         }
 
-        static void DeclareQueue(string address, IModel model, bool durable)
+        void DeclareQueue(string address, IModel model)
         {
-            var arguments = new Dictionary<string, object>
+            if (Address != null && Address.Equals(address))
             {
-                {"x-ha-policy", "all"}
```
Comment (mookid8000) [inline Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs]: I am not an expert on RabbitMQ, but isn't this required for RabbitMQ clustering to work?

File: Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs [hunk 2]
```diff
@@ -447,49 +473,132 @@ CustomQueueingConsumer InitializeConsumer()
...
                     if (_declareInputQueue)
                     {
-                        DeclareQueue(destinationAddress, model, true);
-                    }
-
-                    if (_bindInputQueue)
-                    {
-                        BindInputQueue(destinationAddress, model);
+                        // Checks if queue exists, throws if the queue doesn't exist
+                        model.QueueDeclarePassive(destinationAddress);
```
Comment (mookid8000) [inline Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs]: Does this throw an exception if the queue does not exist? 

We should probably catch & wrap the exception, saying something like "the queue 'dev-null' does not exist"

**Comment by mookid8000** [comment]
Sorry about the late reply 😄 good you made it CLS compliant 👍 I added a couple inline comment. Otherwise really good work, looking forward to getting this merged!

File: Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs (hunk 1 above)
Comment (K3llr) [reply Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs]: Yes, I moved it to the RabbitMqQueueOptionsBuilder as a default setting (with the ability to add/or override arguments.

Comment (mookid8000) [reply Rebus.RabbitMq/RabbitMq/RabbitMqTransport.cs]: ⚡️ 👍

**Comment by amid0** [comment]
Hello. Is the any assumption when wait for release with this fix?

**Comment by mookid8000** [comment]
wow sorry, I completely forgot about this – sorry! 😓 

I have time to look at it tomorrow, so hopefully I can merge it then.

Thanks @amid0 for pinging me! 😅

**Comment by K3llr** [comment]
I also forgot about this for a while. Thanks for doing the small fixes and the housecleaning on the RabbitMqHeaderTest.

**Pull request merged**